File: algo_problems/q161-180/q163.java
class Solution {
    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        long pre = lower;
        pre -= 1;
        List<String> result = new ArrayList();
        for (int num: nums) {
        	if (pre >= num) {
        		continue;
        	}
        	if (pre + 1 != num) {
        		if (num - pre == 2) {
        			result.add(Long.toString(pre + 1));
        		}else{
        			result.add(String.format("%d->%d", pre + 1, num - 1));
        		}
        	}
        	pre = num;
        }
        if (pre != upper) {
    		if (upper - pre == 1) {
    			result.add(Integer.toString(upper));
    		}else{
    			result.add(String.format("%d->%d", pre + 1, upper));
        	}
        }
        return result;
    }
}

File: algo_problems/q161-180/q163.py
class Solution:
    def findMissingRanges(self, nums, lower, upper):
        """
        :type nums: List[int]
        :type lower: int
        :type upper: int
        :rtype: List[str]
        """
        result = []
        if len(nums) == 0:
            return [str(lower)] if lower == upper else ['%d->%d' % (lower, upper)]
        prev = lower - 1
        for i, num in enumerate(nums):
            if num <= prev:
                continue
            if prev != num - 1:
                if num - prev == 2:
                    result.append(str(prev + 1))
                else:
                    result.append("%d->%d" % (prev + 1, num - 1))
            prev = num
        if upper != nums[-1]:
            if upper - nums[-1] == 1:
                result.append(str(upper))
            else:
                result.append("%d->%d" % (nums[-1] + 1, upper))
        return result
